### PR TITLE
[ExtendLifetimes] Add extend lifetimes to emit fake uses from clang

### PR DESCRIPTION
**Pull request by SLTozer**
This patch adds flags to clang to emit fake use intrinsics into IR, preserving the value of variables through codegen to improve the debugging experience. The two flags added are `-fextend-lifetimes`, which extends the lifetime of all variables, and `-fextend-this-ptr`, which extends the lifetime of `this` only. Both of these flags are ~incompatible~ ignored at -O0, since without optimizations there is no purpose to extended variable lifetimes; the `-fextend-lifetimes` flag is intended to be set by default at -Og in a later patch.

Using either of these flags adds the `optdebug` attribute to generated functions, which currently only has the purpose of disabling post-RA machine scheduling, due to its negative effect on variable lifetimes that would have been preserved by fake uses (this behaviour is added in a separate patch).

Original feature and all (or very almost all) code and tests in this patch originally written by @wolfy1961, while I'll be handling reviews and merging.

### CODE REVIEW

File: clang/lib/CodeGen/CGCall.cpp [hunk 1]
```diff
@@ -2543,6 +2543,10 @@ void CodeGenModule::ConstructAttributeList(StringRef Name,
     if (shouldDisableTailCalls())
       FuncAttrs.addAttribute("disable-tail-calls", "true");
 
+    // Suppress the machine instruction scheduler when -fextend-lifetimes is on.
+    if (CodeGenOpts.ExtendLifetimes)
```
Comment (pogo59) [inline clang/lib/CodeGen/CGCall.cpp:2547]: Also -fextend-this-ptr?

File: clang/lib/CodeGen/CGCall.cpp [hunk 2]
```diff
@@ -3558,15 +3562,26 @@ static llvm::StoreInst *findDominatingStoreToReturnValue(CodeGenFunction &CGF) {
     llvm::BasicBlock *IP = CGF.Builder.GetInsertBlock();
     if (IP->empty()) return nullptr;
 
-    // Look at directly preceding instruction, skipping bitcasts and lifetime
-    // markers.
+    // Look at directly preceding instruction, skipping bitcasts, lifetime
+    // markers, and fake uses and their operands.
+    const llvm::Instruction *LoadIntoFakeUse = nullptr;
     for (llvm::Instruction &I : make_range(IP->rbegin(), IP->rend())) {
+      // Ignore instructions are just loads for fake uses; the load should
```
Comment (pogo59) [inline clang/lib/CodeGen/CGCall.cpp:3569]:
```suggestion
      // Ignore instructions that are just loads for fake uses; the load should
```
?

File: clang/lib/CodeGen/CGDecl.cpp [hunk 3]
```diff
@@ -1412,6 +1420,39 @@ void CodeGenFunction::EmitAndRegisterVariableArrayDimensions(
   }
 }
 
+/// Return the maximum size of an aggregate for which we generate a fake use
+/// intrinsic when -fextend-lifetimes is in effect.
+static uint64_t maxFakeUseAggregateSize(const ASTContext &C) {
+  return 4 * C.getTypeSize(C.UnsignedIntTy);
```
Comment (pogo59) [inline clang/lib/CodeGen/CGDecl.cpp:1426]: This limit should be documented in the associated release note.

File: clang/lib/CodeGen/CGDecl.cpp [hunk 4]
```diff
@@ -2523,6 +2575,15 @@ llvm::Function *CodeGenModule::getLLVMLifetimeEndFn() {
   return LifetimeEndFn;
 }
 
+/// Lazily declare the @llvm.fake.use intrinsic.
+llvm::Function *CodeGenModule::getLLVMFakeUseFn() {
+  if (FakeUseFn)
+    return FakeUseFn;
+  FakeUseFn =
+      llvm::Intrinsic::getDeclaration(&getModule(), llvm::Intrinsic::fake_use);
+  return FakeUseFn;
```
Comment (pogo59) [inline clang/lib/CodeGen/CGDecl.cpp:2584]:
```suggestion
  if (!FakeUseFn)
    FakeUseFn =
        llvm::Intrinsic::getDeclaration(&getModule(), llvm::Intrinsic::fake_use);
  return FakeUseFn;
```
This seems a bit simpler to read, but I'm not insisting.

**Review by pogo59 — COMMENTED** [review]
I didn't really look at the tests.

Rather than saying the new options are "incompatible" with -O0, I'd say they are "ignored" at -O0, which is in fact how they are implemented. "Incompatible" suggests to me that there'd be a diagnostic, and there's no need for that.

File: clang/lib/CodeGen/CGDecl.cpp (hunk 3 above)
Comment (SLTozer) [reply clang/lib/CodeGen/CGDecl.cpp:1426]: Good call - the release note also ought to be part of this patch imo.

**Comment by SLTozer** [comment]
Split into two new reviews: https://github.com/llvm/llvm-project/pull/110000 & https://github.com/llvm/llvm-project/pull/110102

**Pull request closed without merging**